### PR TITLE
`A7-3-1`: Improve performance, reduce false positives

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Fixes https://github.com/github/codeql-coding-standards/issues/359 and https://github.com/github/codeql-coding-standards/issues/360.

This pull request addresses a number of issues with `HiddenInheritedNonOverridableMemberFunction.ql`:
   - Addresses potential performance issues and fix false positives related to an incorrect implementation of ordering between `using` declarations and function declarations in a class. We previously required the using declaration to appear first, but order does not matter.
   - Addresses potential performance issues related to matching functions by name.
   - Reduce duplication by reporting `Function`s not `FunctionDeclarationEntry`s in both the hiding and hidden declarations.
   - Exclude special member functions, which cannot be inherited.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `A7-3-1`

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/autosar/test/rules/A7-3-1/test.cpp [hunk 1]
```diff
@@ -60,3 +60,25 @@ void f1() {
...
   l3.f2(l4); // calls C1:f2(S1)
 }
+
+class C5 : public C1 {
+public:
+  void f1(double); // COMPLIANT
+  using C1::f1;    // order of using and f1 declaration is not relevant
+
+  void f2(double) override; // COMPLIANT
+  using C1::f2;             // order of using and f1 declaration is not relevant
```
Comment (rvermeulen) [inline cpp/autosar/test/rules/A7-3-1/test.cpp:70]:
```suggestion
  using C1::f2;             // order of using and f2 declaration is not relevant
```

**Review by rvermeulen — COMMENTED** [review]
Do we want to apply the same optimization for cpp/autosar/src/rules/A7-3-1/HiddenInheritedOverridableMemberFunction.ql?

**Comment by lcartey** [comment]
> Do we want to apply the same optimization for cpp/autosar/src/rules/A7-3-1/HiddenInheritedOverridableMemberFunction.ql?

I updated `HiddenInheritedOverridableMemberFunction.ql` to apply the same logic around `using` declarations - e.g. not to be sensitive to the line location. I did not observe a need for the other optimizations.

File: cpp/autosar/src/rules/A7-3-1/HiddenInheritedNonOverridableMemberFunction.ql [hunk 2]
```diff
@@ -15,24 +15,43 @@
...
-    ude.getDeclaration() = hiddenDecl.getDeclaration() and
-    ude.getEnclosingElement() = overridingDecl.getDeclaration().getDeclaringType() and
-    ude.getLocation().getStartLine() < overridingDecl.getLocation().getStartLine()
+    ude.getDeclaration() = hiddenDecl and
+    ude.getEnclosingElement() = overridingDecl.getDeclaringType()
   ) and
   // Exclude compiler generated member functions which include things like copy constructor that hide base class
   // copy constructors.
-  not overridingDecl.getDeclaration().isCompilerGenerated()
+  not overridingDecl.isCompilerGenerated() and
```
Comment (rvermeulen) [inline cpp/autosar/src/rules/A7-3-1/HiddenInheritedNonOverridableMemberFunction.ql:52]: The OpenPilot results show we are finding issues in macros like `QObject`.
The suggestion removes 38 of those results.

```suggestion
  not overridingDecl.isCompilerGenerated() and
  not overridingDecl.isInMacroExpansion() and
```

**Review by rvermeulen — CHANGES_REQUESTED** [review]
A suggestion to remove results that cannot be acted upon by developers.

File: cpp/autosar/src/rules/A7-3-1/HiddenInheritedNonOverridableMemberFunction.ql (hunk 2 above)
Comment (lcartey) [reply cpp/autosar/src/rules/A7-3-1/HiddenInheritedNonOverridableMemberFunction.ql:54]: I don't think we should generally exclude results for this rule because they were generated by a macro. If I use a macro to generate a member function that hides base class member functions it still contravenes the rule, and addressable by adding a "using ...".

For OpenPilot, QT has a bit of an unusual meta-object system, implemented through the `Q_OBJECT` macro. This macro generates common member functions for QT objects. Reviewing the results for this rule we seemed to be flagging 3 generated functions: `qt_static_metacall`, `tr` and `trUtf8`. `tr` and `trUtf8` seem to be genuine contraventions of this rule - as the macro is generating an implementation of a non-virtual member function that hides the base class implementation.

`qt_static_metacall` is interesting because it's actually a private (static) member function. I don't think we should be reporting _private_ member functions from the base class, because they would not be visible in the subclass anyway, so no hiding is going on. I've therefore updated the query to exclude this case.

**Review by rvermeulen — APPROVED** [review]
We keep contraventions in macro expansions and rely on deviations to exclude those that are caused by patterns outside of the control of the developer (e.g., Qt framework)
